onnx/test/cpp/parser_test.cc

```diff
@@ -0,0 +1,219 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "gtest/gtest.h"
+
+#include "onnx/defs/parser.h"
+#include "onnx/defs/printer.h"
+#include "onnx/checker.h"
+
+using namespace ONNX_NAMESPACE::Utils;
+
+namespace ONNX_NAMESPACE {
+namespace Test {
+
+TEST(ParserTest, TypeTest) {
+  TypeProto type;
+
+  // 1-dimensional tensor type with symbolic dimension:
+  OnnxParser::Parse(type, "float[N]");
+  EXPECT_TRUE(type.has_tensor_type());
+  int float_type = static_cast<int>(TensorProto_DataType::TensorProto_DataType_FLOAT);
+  EXPECT_EQ(type.tensor_type().elem_type(), float_type);
+  EXPECT_TRUE(type.tensor_type().has_shape());
+  EXPECT_EQ(type.tensor_type().shape().dim_size(), 1);
+  EXPECT_EQ(type.tensor_type().shape().dim(0).dim_param(), "N");
+
+  // scalar type:
+  OnnxParser::Parse(type, "float");
+  EXPECT_TRUE(type.has_tensor_type());
+  EXPECT_EQ(type.tensor_type().elem_type(), float_type);
+  EXPECT_TRUE(type.tensor_type().has_shape());
+  EXPECT_EQ(type.tensor_type().shape().dim_size(), 0);
+
+  // tensor type with unknown rank:
+  OnnxParser::Parse(type, "float[]");
+  EXPECT_TRUE(type.has_tensor_type());
+  EXPECT_EQ(type.tensor_type().elem_type(), float_type);
+  EXPECT_FALSE(type.tensor_type().has_shape());
+
+  // 3-dimensional tensor
+  OnnxParser::Parse(type, "float[N,M,K]");
+  EXPECT_EQ(type.tensor_type().shape().dim_size(), 3);
+
+  // Unspecified dimension (neither symbolic nor constant)
+  OnnxParser::Parse(type, "float[N,?,K]");
+  EXPECT_FALSE(type.tensor_type().shape().dim(1).has_dim_param());
+  EXPECT_FALSE(type.tensor_type().shape().dim(1).has_dim_value());
+}
+
+TEST(ParserTest, AttributeTest) {
+  AttributeProto attr;
+
+  OnnxParser::Parse(attr, "x = 2");
+  EXPECT_EQ(attr.name(), "x");
+  EXPECT_EQ(attr.type(), AttributeProto_AttributeType::AttributeProto_AttributeType_INT);
+  EXPECT_EQ(attr.i(), 2);
+
+  OnnxParser::Parse(attr, "x = 0.625");
```

Comment:
curious - will x=0.625f work?

Reply:
Good question. At this point, attributes support only int or float. So, we can use decimal points to distinguish between the two, and I think that is sufficient. (If we ever have multiple precisions for floats in attributes, we could extend this, but I don't see any demand for that now.)

Reply:
But for tensors, it could make sense. Depends on the syntax for tensors (whether the type should be specified separately or not). May be worth thinking about.